Zend/zend_language_parser.y

```diff
@@ -1348,85 +1344,112 @@ isset_variable:
-			toklen = (tok2 - tok1) + 1;
-		} else {
-			tok1 = tok2 = NULL;
-			toklen = 0;
-		}
-
-		if (end == NULL) {
-			len = LANG_SCNG(yy_leng) > 30 ? 30 : LANG_SCNG(yy_leng);
-		} else {
-			len = (end - str) > 30 ? 30 : (end - str);
-		}
-		if (yyres) {
-			if (toklen) {
-				snprintf(buffer, sizeof(buffer), "'%.*s' %.*s", len, str, toklen, tok1);
-			} else {
-				snprintf(buffer, sizeof(buffer), "'%.*s'", len, str);
-			}
-			yystpcpy(yyres, buffer);
-		}
-		return len + (toklen ? toklen + 1 : 0) + 2;
-	}
-
-	/* One of the expected tokens */
-	if (!yyres) {
-		return strlen(yystr) - (*yystr == '"' ? 2 : 0);
-	}
+  return a < b ? a : b;
+}
 
-	if (*yystr == '"') {
-		YYSIZE_T yyn = 0;
-		const char *yyp = yystr;
+struct string {
+  const unsigned char *str;
+  unsigned int len;
+};
+
+// The name of this token.
+// E.g. YYSYMBOL_T_IS_SMALLER_OR_EQUAL => "T_IS_SMALLER_OR_EQUAL"
+// Or null.
+static
+struct string
+token_name (yysymbol_kind_t kind)
+{
+  const char *unexp_str = yysymbol_name (kind);
+  const unsigned int unexp_len = (unsigned int)strlen(unexp_str);
+  const unsigned char *tok1 = memchr(unexp_str, '(', unexp_len);
+  const unsigned char *tok2 = zend_memrchr(unexp_str, ')', unexp_len);
+
+  struct string res;
+  res.str = tok1;
+  res.len = tok1 != NULL && tok2 != NULL ? tok2 - tok1 + 1 : 0;
+  return res;
+}
 
-		for (; *++yyp != '"'; ++yyn) {
-			yyres[yyn] = *yyp;
-		}
-		yyres[yyn] = '\0';
-		return yyn;
-	}
-	yystpcpy(yyres, yystr);
-	return strlen(yystr);
+int
+yyreport_syntax_error (const yypcontext_t *ctx)
+{
+  int res = 0;
+
+  // Report the unexpected token.
+  char unexpected[120];
+  {
+    yysymbol_kind_t unexp = yypcontext_token (ctx);
+    if (unexp == YYSYMBOL_YYEOF) {
+      strncpy(unexpected, yysymbol_name (unexp), sizeof(unexpected));
+    } else if (unexp != YYSYMBOL_YYEMPTY) {
+      // Include the lexeme.
+      const unsigned char *str = LANG_SCNG(yy_text);
+      const unsigned char *end = memchr(str, '\n', LANG_SCNG(yy_leng));
+      const unsigned int len =
+        umin(30, end != NULL ? end - str : LANG_SCNG(yy_leng));
+
+      // Maybe include the token name.
+      struct string token = token_name (unexp);
+
+      if (token.len) {
+        snprintf(unexpected, sizeof(unexpected), "'%.*s' %.*s", len, str, token.len, token.str);
+      } else {
+        snprintf(unexpected, sizeof(unexpected), "'%.*s'", len, str);
+      }
+    }
+  }
+
+  enum { TOKENMAX = 4 };
+  const char *expected[TOKENMAX];
+  unsigned int num_expected = 0;
+  {
+    yysymbol_kind_t exp[TOKENMAX];
+    int n = yypcontext_expected_tokens (ctx, exp, TOKENMAX);
+    if (n < 0)
+      // Forward errors to yyparse.
+      res = n;
+    else
+      {
+        num_expected = n;
+        for (int i = 0; i < num_expected; ++i)
+          expected[i] = yysymbol_name (exp[i]);
+      }
+  }
+
+  char buff[1024];
+  switch (num_expected)
+    {
+    case 0:
+      snprintf(buff, sizeof buff,
```

Comment:
`ZEND_API size_t zend_spprintf(char **message, size_t max_len, const char *format, ...)` can be used to create a emalloced character pointer that can be freed with efree() (From Zend/zend.h)

Reply:
Although using `smart_string_appendl` would also work here, e.g. to loop over `num_expected` times to append `"or "` then `expected[i]` repeatedly)

Reply:
hi @TysonAndre!  Thanks for the review.  Do you think it's better to use an allocated buffer rather than stack space?  WRT looping over "or", the main difference is that with the approach I took, it is possible to internationalize the error messages.  I don't know if you care in fact.